File: backend/security/validation.py
"""
Data validation and sanitization utilities for registration and form submission.
Enforces 25-sentence limit on text fields and prevents dirty data (XSS, SQL injection, etc.)
"""

import re
import html
from typing import Any, Dict, Optional, Tuple

SENTENCE_LIMIT = 25
SENTENCE_REGEX = re.compile(r'[.!?]+')


def count_sentences(text: str) -> int:
    """
    Count sentences in a string.
    A sentence is delimited by . ! or ?
    """
    if not text or not text.strip():
        return 0
    
    matches = SENTENCE_REGEX.findall(text)
    return len(matches)


def exceeds_sentence_limit(text: str, limit: int = SENTENCE_LIMIT) -> bool:
    """Check if text exceeds the sentence limit."""
    return count_sentences(text) > limit


def validate_sentence_limit(
    text: Optional[str],
    field_name: str,
    limit: int = SENTENCE_LIMIT
) -> Tuple[bool, Optional[str]]:
    """
    Validate that a text field does not exceed the sentence limit.
    Returns (valid: bool, error: str or None)
    """
    if not text or not text.strip():
        return True, None
    
    sentences = count_sentences(text)
    if sentences > limit:
        return False, f"{field_name} exceeds {limit} sentence limit (contains {sentences} sentences)."
    
    return True, None


def sanitize_string(value: Any) -> str:
    """
    Sanitize a string to prevent XSS and basic injection attacks.
    - Removes HTML tags and entities
    - Trims whitespace
    - Removes suspicious patterns
    """
    if not value:
        return ''
    
    clean = str(value).strip()
    
    # Decode HTML entities first
    clean = html.unescape(clean)
    
    # Remove HTML tags (even if they sneak through)
    clean = re.sub(r'<[^>]*>', '', clean)
    
    # Remove control characters (except common whitespace)
    clean = re.sub(r'[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]', '', clean)
    
    # Normalize multiple spaces/newlines
    clean = re.sub(r'\s+', ' ', clean)
    
    return clean.strip()


def sanitize_email(email: Optional[str]) -> str:
    """
    Sanitize an email address.
    - Trims whitespace
    - Converts to lowercase
    - Removes suspicious characters
    """
    if not email:
        return ''
    
    clean = str(email).strip().lower()
    
    # Remove HTML tags
    clean = re.sub(r'<[^>]*>', '', clean)
    
    # Only keep alphanumeric, common email chars
    clean = re.sub(r'[^\w.+\-@]', '', clean)
    
    # Limit to reasonable email length
    return clean[:254]


def sanitize_phone(phone: Optional[str]) -> str:
    """
    Sanitize a phone number.
    - Keeps only digits, +, -, (, ), and spaces
    - Trims whitespace
    """
    if not phone:
        return ''
    
    clean = str(phone).strip()
    
    # Remove all non-phone characters except common separators
    clean = re.sub(r'[^\d+()\-\s]', '', clean)
    
    # Remove extra spaces
    clean = re.sub(r'\s+', ' ', clean).strip()
    
    return clean[:20]  # Reasonable phone length


def sanitize_url(url: Optional[str]) -> str:
    """
    Sanitize a URL/link.
    - Trims whitespace
    - Removes dangerous protocols
    """
    if not url:
        return ''
    
    clean = str(url).strip()
    
    # Remove dangerous protocols
    if re.match(r'^(javascript|data|vbscript):', clean, re.IGNORECASE):
        return ''
    
    # Remove HTML tags
    clean = re.sub(r'<[^>]*>', '', clean)
    
    return clean.strip()[:2048]


def validate_and_sanitize_field(
    field_name: str,
    value: Any,
    required: bool = False,
    min_length: int = 0,
    max_length: int = 500,
    max_sentences: int = SENTENCE_LIMIT,
    field_type: str = 'text',
    sanitize: bool = True
) -> Tuple[str, Optional[str]]:
    """
    Validate and sanitize a single form field.
    Returns (sanitized_value: str, error: str or None)
    """
    cleaned = str(value).strip() if value else ''
    
    # Check required
    if required and not cleaned:
        return '', f"{field_name} is required."
    
    if not cleaned:
        return '', None
    
    # Sanitize based on type
    if sanitize:
        if field_type == 'email':
            cleaned = sanitize_email(cleaned)
        elif field_type == 'phone':
            cleaned = sanitize_phone(cleaned)
        elif field_type == 'url':
            cleaned = sanitize_url(cleaned)
        else:
            cleaned = sanitize_string(cleaned)
    
    # Check length
    if len(cleaned) < min_length:
        return '', f"{field_name} must be at least {min_length} characters."
    
    if len(cleaned) > max_length:
        return '', f"{field_name} must not exceed {max_length} characters."
    
    # Check sentence limit for text fields
    if field_type == 'text' and max_sentences:
        sentences = count_sentences(cleaned)
        if sentences > max_sentences:
            return '', f"{field_name} exceeds {max_sentences} sentence limit (contains {sentences} sentences)."
    
    return cleaned, None


def validate_and_sanitize_form(
    form_data: Dict[str, Any],
    field_rules: Dict[str, Dict[str, Any]]
) -> Tuple[Dict[str, Any], Optional[str]]:
    """
    Validate and sanitize an entire form object.
    Returns (sanitized_dict: dict, error: str or None)
    
    If any field fails validation, returns (partial_dict, error_message)
    """
    sanitized = {}
    
    for field_name, value in form_data.items():
        rules = field_rules.get(field_name, {})
        
        if rules:
            cleaned, error = validate_and_sanitize_field(field_name, value, **rules)
            if error:
                return sanitized, error
            sanitized[field_name] = cleaned
        else:
            # Default sanitization for unlisted fields
            sanitized[field_name] = sanitize_string(value)
    
    return sanitized, None


def validate_registration_data(form_data: Dict[str, Any], role: str) -> Tuple[Dict[str, Any], Optional[str]]:
    """
    Validate registration form data specifically.
    This is the main validation entry point for backend registration endpoints.
    
    Returns (sanitized_dict: dict, error: str or None)
    """
    field_rules = {
        'email': {'required': True, 'field_type': 'email', 'max_length': 254},
        'first_name': {'required': True, 'field_type': 'text', 'max_length': 100, 'max_sentences': 3},
        'last_name': {'required': True, 'field_type': 'text', 'max_length': 100, 'max_sentences': 3},
        'middle_name': {'field_type': 'text', 'max_length': 100, 'max_sentences': 3},
        'password': {'required': True, 'min_length': 8, 'max_length': 128},
        
        # Address fields
        'street': {'field_type': 'text', 'max_length': 200, 'max_sentences': 5},
        'barangay': {'field_type': 'text', 'max_length': 100, 'max_sentences': 3},
        'city': {'field_type': 'text', 'max_length': 100, 'max_sentences': 3},
        'province': {'field_type': 'text', 'max_length': 100, 'max_sentences': 3},
        'region': {'field_type': 'text', 'max_length': 100, 'max_sentences': 3},
        
        # Company/HTE fields
        'company_name': {'field_type': 'text', 'max_length': 255, 'max_sentences': 5},
        'company_address': {'field_type': 'text', 'max_length': 500, 'max_sentences': 10},
        'contact_person': {'field_type': 'text', 'max_length': 200, 'max_sentences': 3},
        'contact_phone': {'field_type': 'phone', 'max_length': 20},
        
        # School/Academic fields
        'school_name': {'field_type': 'text', 'max_length': 255, 'max_sentences': 5},
        'campus': {'field_type': 'text', 'max_length': 200, 'max_sentences': 3},
        'course': {'field_type': 'text', 'max_length': 200, 'max_sentences': 3},
        'department': {'field_type': 'text', 'max_length': 200, 'max_sentences': 3},
        
        # Employment fields
        'supervisor_name': {'field_type': 'text', 'max_length': 200, 'max_sentences': 3},
        'position': {'field_type': 'text', 'max_length': 200, 'max_sentences': 3},
        'employee_id': {'field_type': 'text', 'max_length': 50},
        
        # Other optional fields
        'instructor_email': {'field_type': 'email', 'max_length': 254},
        'username': {'field_type': 'text', 'max_length': 150},
        'age': {'field_type': 'text', 'max_length': 3},
        'address': {'field_type': 'text', 'max_length': 500, 'max_sentences': 10},
    }
    
    return validate_and_sanitize_form(form_data, field_rules)

File: src/app/utils/validation.ts
/**
 * Data validation and sanitization utilities for registration and form submission.
 * Enforces 25-sentence limit on text fields and prevents dirty data (XSS, SQL injection, etc.)
 */

const SENTENCE_LIMIT = 25;
const SENTENCE_REGEX = /[.!?]+/g;

/**
 * Count sentences in a string.
 * A sentence is delimited by . ! or ?
 */
export function countSentences(text: string): number {
  if (!text || !text.trim()) return 0;
  const matches = text.match(SENTENCE_REGEX);
  return matches ? matches.length : 0;
}

/**
 * Check if text exceeds the sentence limit.
 */
export function exceedsSentenceLimit(text: string, limit: number = SENTENCE_LIMIT): boolean {
  return countSentences(text) > limit;
}

/**
 * Validate that a text field does not exceed the sentence limit.
 * Returns { valid: boolean, error?: string }
 */
export function validateSentenceLimit(
  text: string | undefined,
  fieldName: string,
  limit: number = SENTENCE_LIMIT
): { valid: boolean; error?: string } {
  if (!text || !text.trim()) return { valid: true };
  
  const sentences = countSentences(text);
  if (sentences > limit) {
    return {
      valid: false,
      error: `${fieldName} exceeds ${limit} sentence limit (contains ${sentences} sentences).`,
    };
  }
  return { valid: true };
}

/**
 * Sanitize a string to prevent XSS and basic injection attacks.
 * - Removes HTML tags and entities
 * - Trims whitespace
 * - Removes suspicious patterns
 */
export function sanitizeString(value: string | undefined): string {
  if (!value) return '';
  
  let clean = String(value).trim();
  
  // Remove HTML tags
  clean = clean.replace(/<[^>]*>/g, '');
  
  // Decode HTML entities
  const textarea = document.createElement('textarea');
  textarea.innerHTML = clean;
  clean = textarea.value;
  
  // Remove control characters (except common whitespace)
  clean = clean.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g, '');
  
  // Normalize multiple spaces/newlines
  clean = clean.replace(/\s+/g, ' ');
  
  return clean.trim();
}

/**
 * Sanitize an email address.
 * - Trims whitespace
 * - Converts to lowercase
 * - Removes suspicious characters outside standard email format
 */
export function sanitizeEmail(email: string | undefined): string {
  if (!email) return '';
  
  let clean = String(email).trim().toLowerCase();
  
  // Remove leading/trailing whitespace
  clean = clean.trim();
  
  // Remove any HTML tags
  clean = clean.replace(/<[^>]*>/g, '');
  
  // Only keep alphanumeric, common email chars, and one @
  // This is a basic check; full validation done on backend
  clean = clean.replace(/[^\w.+-@]/g, '');
  
  return clean;
}

/**
 * Sanitize a phone number.
 * - Keeps only digits, +, -, (, )
 * - Trims whitespace
 */
export function sanitizePhone(phone: string | undefined): string {
  if (!phone) return '';
  
  let clean = String(phone).trim();
  
  // Remove all non-phone characters except common separators
  clean = clean.replace(/[^\d+()\-\s]/g, '');
  
  // Remove extra spaces
  clean = clean.replace(/\s+/g, ' ').trim();
  
  return clean;
}

/**
 * Sanitize a URL/link.
 * - Trims whitespace
 * - Ensures it starts with http:// or https://
 * - Removes javascript: and data: protocols
 */
export function sanitizeUrl(url: string | undefined): string {
  if (!url) return '';
  
  let clean = String(url).trim();
  
  // Remove dangerous protocols
  if (clean.toLowerCase().match(/^(javascript|data|vbscript):/)) {
    return '';
  }
  
  // Remove HTML tags
  clean = clean.replace(/<[^>]*>/g, '');
  
  // Ensure http/https if it looks like a URL
  if (clean && !clean.match(/^https?:\/\//)) {
    if (clean.includes('.')) {
      clean = 'https://' + clean;
    }
  }
  
  return clean.trim();
}

/**
 * Validate and sanitize a single form field.
 * Returns the sanitized value if valid, or throws an error.
 */
export function validateAndSanitizeField(
  fieldName: string,
  value: string | undefined,
  options: {
    required?: boolean;
    minLength?: number;
    maxLength?: number;
    maxSentences?: number;
    type?: 'text' | 'email' | 'phone' | 'url';
    sanitize?: boolean;
  } = {}
): string {
  const {
    required = false,
    minLength = 0,
    maxLength = 500,
    maxSentences = SENTENCE_LIMIT,
    type = 'text',
    sanitize = true,
  } = options;
  
  let cleaned = value ? String(value).trim() : '';
  
  // Check required
  if (required && !cleaned) {
    throw new Error(`${fieldName} is required.`);
  }
  
  if (!cleaned) return '';
  
  // Sanitize based on type
  if (sanitize) {
    switch (type) {
      case 'email':
        cleaned = sanitizeEmail(cleaned);
        break;
      case 'phone':
        cleaned = sanitizePhone(cleaned);
        break;
      case 'url':
        cleaned = sanitizeUrl(cleaned);
        break;
      default:
        cleaned = sanitizeString(cleaned);
    }
  }
  
  // Check length
  if (cleaned.length < minLength) {
    throw new Error(`${fieldName} must be at least ${minLength} characters.`);
  }
  
  if (cleaned.length > maxLength) {
    throw new Error(`${fieldName} must not exceed ${maxLength} characters.`);
  }
  
  // Check sentence limit for text fields
  if (type === 'text' && maxSentences) {
    const sentences = countSentences(cleaned);
    if (sentences > maxSentences) {
      throw new Error(`${fieldName} exceeds ${maxSentences} sentence limit (contains ${sentences} sentences).`);
    }
  }
  
  return cleaned;
}

/**
 * Validate and sanitize an entire form object.
 * Returns a sanitized copy of the form data.
 * Throws an error if any field fails validation.
 */
export function validateAndSanitizeForm(
  formData: Record<string, any>,
  fieldRules: Record<string, Parameters<typeof validateAndSanitizeField>[2]>
): Record<string, any> {
  const sanitized: Record<string, any> = {};
  
  for (const [fieldName, value] of Object.entries(formData)) {
    const rules = fieldRules[fieldName];
    
    if (rules) {
      sanitized[fieldName] = validateAndSanitizeField(fieldName, value, rules);
    } else {
      // Default sanitization for unlisted fields
      sanitized[fieldName] = sanitizeString(value);
    }
  }
  
  return sanitized;
}

/**
 * Validate registration form data specifically.
 * This is the main validation entry point for the Register page.
 */
export function validateRegistrationData(
  formData: Record<string, any>,
  role: string
): Record<string, any> {
  const fieldRules: Record<string, Parameters<typeof validateAndSanitizeField>[2]> = {
    email: { required: true, type: 'email', maxLength: 254 },
    firstName: { required: true, type: 'text', maxLength: 100, maxSentences: 3 },
    lastName: { required: true, type: 'text', maxLength: 100, maxSentences: 3 },
    middleInitial: { type: 'text', maxLength: 10 },
    name: { type: 'text', maxLength: 255, maxSentences: 3 },
    password: { required: true, minLength: 8, maxLength: 128 },
    confirmPassword: { required: true, minLength: 8, maxLength: 128 },
    
    // Address fields
    street: { type: 'text', maxLength: 200, maxSentences: 5 },
    barangay: { type: 'text', maxLength: 100, maxSentences: 3 },
    barangayManual: { type: 'text', maxLength: 100, maxSentences: 3 },
    city: { type: 'text', maxLength: 100, maxSentences: 3 },
    province: { type: 'text', maxLength: 100, maxSentences: 3 },
    region: { type: 'text', maxLength: 100, maxSentences: 3 },
    
    // Company/HTE fields
    companyName: { type: 'text', maxLength: 255, maxSentences: 5 },
    companyAddress: { type: 'text', maxLength: 500, maxSentences: 10 },
    contactPerson: { type: 'text', maxLength: 200, maxSentences: 3 },
    contactPhone: { type: 'phone', maxLength: 20 },
    
    // School/Academic fields
    schoolName: { type: 'text', maxLength: 255, maxSentences: 5 },
    campus: { type: 'text', maxLength: 200, maxSentences: 3 },
    course: { type: 'text', maxLength: 200, maxSentences: 3 },
    department: { type: 'text', maxLength: 200, maxSentences: 3 },
    
    // Employment fields
    supervisorName: { type: 'text', maxLength: 200, maxSentences: 3 },
    position: { type: 'text', maxLength: 200, maxSentences: 3 },
    employeeId: { type: 'text', maxLength: 50 },
    
    // Other optional fields
    instructorEmail: { type: 'email', maxLength: 254 },
    username: { type: 'text', maxLength: 150 },
  };
  
  return validateAndSanitizeForm(formData, fieldRules);
}
